### DISCUSSION TITLE
Defining default modulators

### DISCUSSION THREAD

**Discussion by davy7125** [SFe-Team-was-taken/SFE#38] [Ideas]
Having read the DMOD specifications of this [page](https://github.com/spessasus/soundfont-proposals/blob/main/default_modulators.md), I was wondering if it wouldn't be simpler to consider the DMOD section as a complete definition of the default modulators?

Currently this section is considered to be an override of the default modulators already specified in the 2.04 specifications: the default modulators should be applied first, then those of the DMOD section.

The problem is that the default modulators slightly changed between 2.01 and 2.04 (velocity to cut off frequency) so even with the DMOD section there is room for a small ambiguity.

To the contrary, we could consider that if at least one default modulator is set in the DMOD section, all default modulators coming from the 2.04 specifications will be ignored. In this case, we know exactly what is happening. The only drawback I see is when we want NO default modulators.

An alternative can also be: if the DMOD section is in the file, may it be empty, then we ignore all default modulators.

**Comment by sylvia-leaf** [comment]
### `DMOD` vs legacy SF 2.0x default modulators

This is a good idea, however work on the `DMOD` sub-chunk is not immediately necessary. This is because for SFe 4.0, which doesn't include the `DMOD` sub-chunk to avoid "scope creep", the default modulators remain the same as legacy SF 2.04. The 2.01 modulators are not used in SFe 4 unless defined by `DMOD` (only available in 4.1 and later), because the velocity-to-cutoff modulator change seems to be a bugfix.

This begs the question: how many banks do we really think rely on the quirks of the 2.01 specification in particular? And even if they do exist, we can define the 2.01 version of that modulator using the future `DMOD` sub-chunk.

However, my suggestion is that we use *no* default modulators if an SFe version of 4.1 or later is detected under any circumstances. If the `DMOD` sub-chunk doesn't contain any records, there are no default modulators. To make it easier for legacy SF bank developers to create new SFe banks, SFe editors such as the planned Polyphone 3 would add the 2.04 default modulators to the `DMOD` sub-chunk when a legacy 2.0x or SFe 4.0 bank is opened and saved as SFe 4.1 or later. An SFe editor could also include an option to enable or disable automatic creation of a `DMOD` sub-chunk containing the 2.04 default modulators.

### `DMOD` implementation in SFe 4.1

The `DMOD` implementation in SFe will have slightly more information than the original `DMOD` sub-chunk definition, as you have pointed out that the linked `DMOD` specification could have a few things added to it to describe how implementations should interpret the `DMOD` sub-chunk.

Solving the implied default modulator problem starts with the `DMOD` sub-chunk, however we'll use other sub-chunks to define important information. We think that we can add more defined values for the `SFty` sub-chunk, for example `SFe-static with 2.04 modulators` and/or `SFe-static with 2.01 modulators`.

However, one of the most important things to know is that the `DMOD` implementation in SFe 4.1 won't be compatible with Spessasynth's implementation out of the box. This is because the `DMOD` (and `PNMM`) sub-chunks will also be moved from the root of the `INFO-list` chunk to the `ISFe-list` sub-chunk to make it clearer that these are SFe features and aren't available in SF2.04. While it won't be compatible out of the box, it should not be a complete rewrite, because the `ISFe-list` sub-chunk will be required for SFe support anyway. However, SFe players would likely also need to be able to read `DMOD` sub-chunks in the root of the `INFO-list` chunk, complicating things. If we can agree on where the `DMOD` sub-chunk goes (for example `DMOD` is always found inside of `ISFe-list`), then we can reduce required complexity for SFe implementations.

### What to do next

We can continue discussion on how we will implement custom default modulators and/or the `DMOD` sub-chunk, because these features are not planned for SFe 4.0. Therefore, we have a lot of time to come up with a workable solution in preparation for SFe 4.1.

**Comment by spessasus** [comment]
I've updated the DMOD proposal.
The DMOD chunk now replaces the default modulator list.
